client/document.go

```diff
@@ -119,15 +122,250 @@ func NewDocFromMap(data map[string]any) (*Document, error) {
+
+	case FieldKind_DATETIME:
+		return getDateTime(val)
+
+	case FieldKind_INT:
+		return getInt64(val)
+
+	case FieldKind_INT_ARRAY:
+		return getArray(val, getInt64)
+
+	case FieldKind_NILLABLE_INT_ARRAY:
+		return getNillableArray(val, getInt64)
+
+	case FieldKind_FOREIGN_OBJECT:
+		return getString(val)
+
+	case FieldKind_FOREIGN_OBJECT_ARRAY:
+		return nil, NewErrFieldOrAliasToFieldNotExist(field.Name)
+	}
+
+	return nil, NewErrUnhandledType("FieldKind", field.Kind)
+}
+
+func getString(v any) (string, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		b, err := val.StringBytes()
+		return string(b), err
+	default:
+		return val.(string), nil
+	}
+}
+
+func getBool(v any) (bool, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		return val.Bool()
+	default:
+		return val.(bool), nil
+	}
+}
+
+func getFloat64(v any) (float64, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		return val.Float64()
+	default:
+		switch v := val.(type) {
+		case int:
+			return float64(v), nil
+		case int64:
+			return float64(v), nil
+		case float64:
+			return v, nil
+		default:
+			return 0, NewErrUnexpectedType[float64]("field", v)
+		}
+	}
+}
+
+func getInt64(v any) (int64, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		return val.Int64()
+	default:
+		switch v := val.(type) {
+		case int:
+			return int64(v), nil
+		case int64:
+			return v, nil
+		case float64:
+			return int64(v), nil
+		default:
+			return 0, NewErrUnexpectedType[int64]("field", v)
+		}
+	}
+}
+
+func getDateTime(v any) (time.Time, error) {
+	var s string
+	switch val := v.(type) {
+	case *fastjson.Value:
+		b, err := val.StringBytes()
+		if err != nil {
+			return time.Time{}, err
+		}
+		s = string(b)
+	default:
+		s = val.(string)
+	}
+	return time.Parse(time.RFC3339, s)
+}
+
+func getArray[T any](
+	v any,
+	typeGetter func(any) (T, error),
+) ([]T, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		if val.Type() == fastjson.TypeNull {
+			return nil, nil
+		}
+
+		valArray, err := val.Array()
+		if err != nil {
+			return nil, err
+		}
+
+		arr := make([]T, len(valArray))
+		for i, arrItem := range valArray {
+			if arrItem.Type() == fastjson.TypeNull {
+				continue
+			}
+			arr[i], err = typeGetter(arrItem)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return arr, nil
+	default:
+		return val.([]T), nil
+	}
+}
+
+func getNillableArray[T any](
+	v any,
+	typeGetter func(any) (T, error),
+) ([]*T, error) {
+	switch val := v.(type) {
+	case *fastjson.Value:
+		if val.Type() == fastjson.TypeNull {
+			return nil, nil
+		}
+
+		valArray, err := val.Array()
+		if err != nil {
+			return nil, err
+		}
+
+		arr := make([]*T, len(valArray))
+		for i, arrItem := range valArray {
+			if arrItem.Type() == fastjson.TypeNull {
+				continue
+			}
+			v, err := typeGetter(arrItem)
+			if err != nil {
+				return nil, err
+			}
+			arr[i] = &v
+		}
+
+		return arr, nil
+	default:
+		return val.([]*T), nil
```

Comment:
issue: this could panic if the types do not match

Reply:
✅